working/macros/example/data_class.dart

```diff
@@ -5,225 +5,235 @@ import '../api/code.dart';
-class _CopyWith implements ClassMacro {
+class _CopyWith implements ClassDeclarationsMacro {
   const _CopyWith();
 
   @override
-  void visitClass(ClassDeclaration clazz, ClassContext context) {
-    context.buildDeclarations((builder) async {
-      var methods = await builder.methodsOf(clazz);
-      if (methods.any((c) => c.name == 'copyWith')) {
-        throw ArgumentError(
-            'Cannot generate a copyWith method because one already exists');
-      }
-      var allFields = await clazz.allFields(builder).toList();
-      var namedParams = [
-        for (var field in allFields)
-          ParameterCode.fromString(
-              '${field.type.toCode()}${field.type.isNullable ? '' : '?'} '
-              '${field.name}'),
-      ];
-      var args = [
-        for (var field in allFields)
-          NamedArgumentCode.fromString(
-              '${field.name}: ${field.name} ?? this.${field.name}'),
-      ];
-      builder.declareInClass(DeclarationCode.fromParts([
-        clazz.instantiate().toCode(),
-        ' copyWith({',
-        ...namedParams.joinAsCode(', '),
-        ',})',
-        // TODO: We assume this constructor exists, but should check
-        '=> ', clazz.instantiate().toCode(), '(',
-        ...args.joinAsCode(', '),
-        ', );',
-      ]));
-    });
+  Future<void> buildDeclarationsForClass(
+      ClassDeclaration clazz, ClassDeclarationBuilder builder) async {
+    var methods = await builder.methodsOf(clazz);
+    if (methods.any((c) => c.name == 'copyWith')) {
+      throw ArgumentError(
+          'Cannot generate a copyWith method because one already exists');
+    }
+    var allFields = await clazz.allFields(builder).toList();
+    var namedParams = [
+      for (var field in allFields)
+        ParameterCode.fromString(
+            '${field.type.toCode()}${field.type.isNullable ? '' : '?'} '
+            '${field.name}'),
+    ];
+    var args = [
+      for (var field in allFields)
+        NamedArgumentCode.fromString(
+            '${field.name}: ${field.name} ?? this.${field.name}'),
+    ];
+    builder.declareInClass(DeclarationCode.fromParts([
+      clazz.instantiate().toCode(),
+      ' copyWith({',
+      ...namedParams.joinAsCode(', '),
+      ',})',
+      // TODO: We assume this constructor exists, but should check
+      '=> ', clazz.instantiate().toCode(), '(',
+      ...args.joinAsCode(', '),
+      ', );',
+    ]));
   }
 }
 
 const hashCode = _HashCode();
 
-class _HashCode implements ClassMacro {
+class _HashCode implements ClassDeclarationsMacro, ClassDefinitionMacro {
   const _HashCode();
 
   @override
-  void visitClass(ClassDeclaration clazz, ClassContext context) {
-    context.buildDeclarations((builder) {
-      builder.declareInClass(DeclarationCode.fromString('''
+  void buildDeclarationsForClass(
+      ClassDeclaration clazz, ClassDeclarationBuilder builder) {
+    builder.declareInClass(DeclarationCode.fromString('''
 @override
 external int get hashCode;'''));
```

Comment:
Say a user wrote:

```dart
class _HashCode implements ClassDeclarationsMacro {
  const _HashCode();

  @override
  void buildDeclarationsForClass(
      ClassDeclaration clazz, ClassDeclarationBuilder builder) {
    var hashCodeExprs = [
      await for (var field in clazz.allFields(builder))
        ExpressionCode.fromString('${field.name}.hashCode')
    ].joinAsCode(' ^ ');

    builder.declareInClass(DeclarationCode.fromParts([
      '@override int get hashCode =>',
      ...hashCodeExprs,
      ';']));
  }
}
```

So they both declare and define the method all in the declaration phase. The API will let them do this, and it will work correctly in cases where there are no other macros adding fields. But if that isn't the case this simpler code will silently do the wrong thing. Do I have that right?

If we going to foist the complexity of explicit phased interfaces onto users, it would be really nice to at least protect them from misusing the phases. I'm not necessarily saying we should go back to an explicitly phased introspection API because that was a *lot* of classes. But this feels like a footgun to me.

Reply:
Ya I think we should continue discussion on this but I am on the fence right now. It is both more convenient and more efficient to declare the body as well, if you have enough information to do so.